lib/internal/modules/cjs/loader.js

```diff
@@ -1529,6 +1560,113 @@ function getMaybeCachedSource(mod, filename) {
   return content;
 }
 
+/**
+ * Resolve and evaluate it synchronously as ESM if it's ESM.
+ * @param {Module} mod CJS module instance
+ * @param {string} filename Absolute path of the file.
+ */
+function loadMTS(mod, filename) {
+  if (StringPrototypeIncludes(filename, 'node_modules')) {
+    throw new ERR_UNSUPPORTED_NODE_MODULES_TYPE_STRIPPING(filename);
+  }
+  const source = getMaybeCachedSource(mod, filename);
+  const { tsParse } = require('internal/modules/helpers');
+  const content = tsParse(source);
+  const cascadedLoader = require('internal/modules/esm/loader').getOrInitializeCascadedLoader();
+  const isMain = mod[kIsMainSymbol];
+  if (isMain) {
+    require('internal/modules/run_main').runEntryPointWithESMLoader((cascadedLoader) => {
+      const mainURL = pathToFileURL(filename).href;
+      cascadedLoader.import(mainURL, undefined, { __proto__: null }, true);
+    });
+    // ESM won't be accessible via process.mainModule.
+    setOwnProperty(process, 'mainModule', undefined);
+  } else {
+    const {
+      wrap,
+      namespace,
+    } = cascadedLoader.importSyncForRequire(mod, filename, content, isMain, mod[kModuleParent]);
+    if (!ObjectHasOwn(namespace, 'default') || ObjectHasOwn(namespace, '__esModule')) {
+      mod.exports = namespace;
+    } else {
+      mod.exports = createRequiredModuleFacade(wrap);
+    }
+  }
```

Comment:
This seems copy-pasted from `loadESMFromCJS`, which is a very new function (added with `require(esm)` and still undergoing active development. I think we should avoid the copy-paste and find a way to share the logic so that as `require(esm)` improves, this copy isn’t left behind.

Reply:
refactored this part